File: packages/sdk/src/desktop.ts
/**
 * `sandbox.desktop` - control a GUI desktop inside a sandbox.
 *
 * Requires a desktop-enabled template (one whose image runs an X server and
 * exposes boxd's DesktopService). Every action is a single authenticated
 * data-plane RPC: a coordinate click is one round trip, a drag is one batched
 * request, and typing has no artificial per-character pacing.
 *
 * A paused sandbox is transparently resumed on first use, like `commands`
 * and `files`.
 *
 * Accessed as `sandbox.desktop.screenshot()`, `sandbox.desktop.click(x, y)`, …
 */

import { dataPlaneTarget } from "./config.js"
import { request } from "./http.js"
import { withTokenRetry } from "./tokenRetry.js"

/** @internal */
export interface DesktopDeps {
  sandboxId: string
  sandboxHost: string
  getAccessToken: () => string
  refreshActivate: () => Promise<string>
  /** Publish the noVNC port and build its public URL (from the Sandbox). */
  publishStreamPort: () => Promise<void>
  streamBaseUrl: () => string
}

const RPC_BASE = "/superserve.boxd.v1.DesktopService"

/** Port the desktop template serves noVNC (websockify) on. */
export const DESKTOP_STREAM_PORT = 6080

/**
 * Cap on a buffered screenshot RPC response. boxd caps a frame at 32 MiB of
 * PNG; base64 in the JSON envelope inflates that by 4/3.
 */
export const MAX_SCREENSHOT_RESPONSE_BYTES = 48 * 1024 * 1024

export type MouseButton = "left" | "right" | "middle"

/** One step of a `desktop.actions()` batch. */
export type DesktopAction =
  | { type: "click"; x: number; y: number; button?: MouseButton }
  | { type: "doubleClick"; x: number; y: number }
  | { type: "move"; x: number; y: number }
  | { type: "mouseDown"; x: number; y: number; button?: MouseButton }
  | { type: "mouseUp"; x: number; y: number; button?: MouseButton }
  | { type: "press"; key: string | string[] }
  | { type: "write"; text: string }
  | { type: "scroll"; dx?: number; dy?: number }

export interface Screenshot {
  /** PNG bytes. */
  data: Uint8Array
  width: number
  height: number
}

export interface StreamUrlOptions {
  /**
   * Open the viewer read-only (viewer-side flag on the noVNC client;
   * interactive control stays available to anyone else with the URL).
   */
  viewOnly?: boolean
}

/**
 * Friendly key names → X keysyms, so agent-facing code can say "enter" or
 * "ctrl". Unlisted names pass through verbatim (all X keysyms stay usable).
 */
const KEYSYMS: Record<string, string> = {
  enter: "Return",
  return: "Return",
  esc: "Escape",
  escape: "Escape",
  tab: "Tab",
  space: "space",
  backspace: "BackSpace",
  delete: "Delete",
  insert: "Insert",
  up: "Up",
  down: "Down",
  left: "Left",
  right: "Right",
  home: "Home",
  end: "End",
  pageup: "Page_Up",
  pagedown: "Page_Down",
  ctrl: "ctrl",
  control: "ctrl",
  alt: "alt",
  shift: "shift",
  cmd: "super",
  win: "super",
  super: "super",
}

function toKeysym(key: string): string {
  return KEYSYMS[key.toLowerCase()] ?? key
}

/**
 * Split a chord into (modifiers, final key) for the KeyEvent RPC.
 * `press("ctrl+c")` and `press(["ctrl", "c"])` are equivalent.
 */
function chordParts(key: string | string[]): {
  modifiers: string[]
  key: string
} {
  const parts = (Array.isArray(key) ? key : key.split("+")).map(toKeysym)
  if (parts.length === 0 || parts.some((p) => p === "")) {
    throw new Error("press: empty key")
  }
  return { modifiers: parts.slice(0, -1), key: parts[parts.length - 1] }
}

function buttonEnum(button: MouseButton = "left"): string {
  switch (button) {
    case "left":
      return "POINTER_BUTTON_LEFT"
    case "right":
      return "POINTER_BUTTON_RIGHT"
    case "middle":
      return "POINTER_BUTTON_MIDDLE"
  }
}

function assertCoordinate(x: number, y: number): void {
  // 0 is a valid coordinate — validate with integer checks, never truthiness.
  if (!Number.isInteger(x) || !Number.isInteger(y) || x < 0 || y < 0) {
    throw new Error(`Invalid coordinates (${x}, ${y}): must be integers >= 0`)
  }
}

/** Proto-JSON shapes for the DesktopService RPCs. @internal */
type PointerEventBody = {
  x: number
  y: number
  button?: string
  action: string
}
type KeyEventBody = { key?: string; text?: string; modifiers?: string[] }
type ScrollEventBody = { dx?: number; dy?: number }
type ActionBody =
  | { pointer: PointerEventBody }
  | { key: KeyEventBody }
  | { scroll: ScrollEventBody }

function pointerBody(
  x: number,
  y: number,
  action: string,
  button?: MouseButton,
): PointerEventBody {
  assertCoordinate(x, y)
  return { x, y, button: buttonEnum(button), action }
}

function actionBody(action: DesktopAction): ActionBody {
  switch (action.type) {
    case "click":
      return {
        pointer: pointerBody(
          action.x,
          action.y,
          "POINTER_ACTION_CLICK",
          action.button,
        ),
      }
    case "doubleClick":
      return {
        pointer: pointerBody(action.x, action.y, "POINTER_ACTION_DOUBLE_CLICK"),
      }
    case "move":
      return {
        pointer: pointerBody(action.x, action.y, "POINTER_ACTION_MOVE"),
      }
    case "mouseDown":
      return {
        pointer: pointerBody(
          action.x,
          action.y,
          "POINTER_ACTION_DOWN",
          action.button,
        ),
      }
    case "mouseUp":
      return {
        pointer: pointerBody(
          action.x,
          action.y,
          "POINTER_ACTION_UP",
          action.button,
        ),
      }
    case "press": {
      const { modifiers, key } = chordParts(action.key)
      return { key: { key, modifiers } }
    }
    case "write":
      return { key: { text: action.text } }
    case "scroll":
      return { scroll: { dx: action.dx ?? 0, dy: action.dy ?? 0 } }
  }
}

export class Desktop {
  private readonly _dataPlaneBaseUrl: string
  private readonly _routingHeaders: Record<string, string>

  /** @internal */
  constructor(private readonly _deps: DesktopDeps) {
    const target = dataPlaneTarget(_deps.sandboxId, _deps.sandboxHost)
    this._dataPlaneBaseUrl = target.url
    this._routingHeaders = target.headers
  }

  /**
   * Capture the current screen as a PNG.
   *
   * One round trip; safe to call while a live stream viewer is open.
   *
   * @example
   * ```typescript
   * const shot = await sandbox.desktop.screenshot()
   * await fs.writeFile("screen.png", shot.data)
   * ```
   */
  async screenshot(): Promise<Screenshot> {
    const raw = await this._rpc<{
      image?: string
      width?: number
      height?: number
    }>("Screenshot", {}, { maxBytes: MAX_SCREENSHOT_RESPONSE_BYTES })
    if (raw.image === undefined) {
      throw new Error("Screenshot response missing image data")
    }
    return {
      data: Uint8Array.from(atob(raw.image), (c) => c.charCodeAt(0)),
      width: raw.width ?? 0,
      height: raw.height ?? 0,
    }
  }

  /** Click at (x, y). One RPC — move and click are a single action. */
  async click(
    x: number,
    y: number,
    options: { button?: MouseButton } = {},
  ): Promise<void> {
    await this._pointer(
      pointerBody(x, y, "POINTER_ACTION_CLICK", options.button),
    )
  }

  /** Double-click at (x, y). */
  async doubleClick(x: number, y: number): Promise<void> {
    await this._pointer(pointerBody(x, y, "POINTER_ACTION_DOUBLE_CLICK"))
  }

  /** Right-click at (x, y). */
  rightClick(x: number, y: number): Promise<void> {
    return this.click(x, y, { button: "right" })
  }

  /** Middle-click at (x, y). */
  middleClick(x: number, y: number): Promise<void> {
    return this.click(x, y, { button: "middle" })
  }

  /** Move the pointer to (x, y) without clicking. */
  async moveMouse(x: number, y: number): Promise<void> {
    await this._pointer(pointerBody(x, y, "POINTER_ACTION_MOVE"))
  }

  /**
   * Drag from one point to another. Executes as one atomic batch
   * (down → move → up) under the sandbox's input lock, so no other input
   * can interleave mid-drag.
   */
  async drag(
    from: [number, number],
    to: [number, number],
    options: { button?: MouseButton } = {},
  ): Promise<void> {
    await this.actions([
      { type: "mouseDown", x: from[0], y: from[1], button: options.button },
      { type: "move", x: to[0], y: to[1] },
      { type: "mouseUp", x: to[0], y: to[1], button: options.button },
    ])
  }

  /**
   * Scroll the viewport under the pointer. Positive `dy` scrolls down,
   * negative up; `dx` scrolls horizontally. Both axes in one call.
   */
  async scroll(options: { dx?: number; dy?: number }): Promise<void> {
    await this._rpc("Scroll", { dx: options.dx ?? 0, dy: options.dy ?? 0 })
  }

  /**
   * Type literal text. Fast by default — no per-character pacing, so long
   * strings land in well under a second.
   */
  async write(text: string): Promise<void> {
    if (text.length === 0) return
    await this._rpc("SendKey", { text })
  }

  /**
   * Press a key or chord: `press("enter")`, `press("ctrl+c")`,
   * `press(["ctrl", "shift", "p"])`. Friendly names (enter, esc, cmd, …)
   * map to X keysyms; unrecognized names pass through verbatim.
   */
  async press(key: string | string[]): Promise<void> {
    const { modifiers, key: finalKey } = chordParts(key)
    await this._rpc("SendKey", { key: finalKey, modifiers })
  }

  /**
   * Execute an ordered batch of actions in a single request.
   *
   * The whole batch is validated before anything runs, then executed under
   * the sandbox's input lock — no other input can interleave. Execution
   * stops at the first failing action. This is the fast path for models
   * that emit several actions per turn.
   *
   * @example
   * ```typescript
   * await sandbox.desktop.actions([
   *   { type: "click", x: 640, y: 32 },
   *   { type: "write", text: "https://example.com" },
   *   { type: "press", key: "enter" },
   * ])
   * ```
   */
  async actions(actions: DesktopAction[]): Promise<void> {
    if (actions.length === 0) return
    await this._rpc("SendActions", { actions: actions.map(actionBody) })
  }

  /**
   * Resize the virtual display. Width must be a multiple of 8 between 320
   * and 8192; height between 200 and 8192. Takes effect live — no restart.
   */
  async resize(width: number, height: number): Promise<void> {
    await this._rpc("Resize", { width, height })
  }

  /**
   * Publish the live desktop viewer (noVNC) and return its browser URL.
   *
   * The URL goes through the sandbox's preview-port access policy — under a
   * private policy, viewers also need a preview token.
   */
  async getStreamUrl(options: StreamUrlOptions = {}): Promise<string> {
    await this._deps.publishStreamPort()
    const base = this._deps.streamBaseUrl()
    const params = new URLSearchParams({ autoconnect: "1", resize: "scale" })
    if (options.viewOnly) params.set("view_only", "1")
    return `${base}/vnc.html?${params.toString()}`
  }

  private async _pointer(body: PointerEventBody): Promise<void> {
    await this._rpc("SendPointer", body)
  }

  /**
   * One unary DesktopService call, Connect JSON protocol: a plain POST of
   * proto-JSON to the procedure path. Paused sandboxes resume via the shared
   * token-retry path.
   */
  private async _rpc<T = Record<string, never>>(
    method: string,
    body: unknown,
    opts: { maxBytes?: number } = {},
  ): Promise<T> {
    const send = (token: string) =>
      request<T>({
        method: "POST",
        url: `${this._dataPlaneBaseUrl}${RPC_BASE}/${method}`,
        headers: { ...this._routingHeaders, "X-Access-Token": token },
        body,
        maxBytes: opts.maxBytes,
      })
    return withTokenRetry(this._deps, send)
  }
}

File: packages/python-sdk/src/superserve/desktop.py
"""``sandbox.desktop`` - control a GUI desktop inside a sandbox.

Requires a desktop-enabled template (one whose image runs an X server and
exposes boxd's DesktopService). Every action is a single authenticated
data-plane RPC: a coordinate click is one round trip, a drag is one batched
request, and typing has no artificial per-character pacing.

A paused sandbox is transparently resumed on first use, like ``commands``
and ``files``.
"""

from __future__ import annotations

import base64
from collections.abc import Awaitable, Callable, Sequence
from dataclasses import dataclass
from typing import Any, Literal
from urllib.parse import urlencode

import httpx

from ._config import data_plane_target
from ._http import api_request, async_api_request
from ._token_retry import async_with_token_retry, with_token_retry

_RPC_BASE = "/superserve.boxd.v1.DesktopService"

#: Port the desktop template serves noVNC (websockify) on.
DESKTOP_STREAM_PORT = 6080

MouseButton = Literal["left", "right", "middle"]

_BUTTONS: dict[str, str] = {
    "left": "POINTER_BUTTON_LEFT",
    "right": "POINTER_BUTTON_RIGHT",
    "middle": "POINTER_BUTTON_MIDDLE",
}

# Friendly key names -> X keysyms, so agent-facing code can say "enter" or
# "ctrl". Unlisted names pass through verbatim (all X keysyms stay usable).
_KEYSYMS: dict[str, str] = {
    "enter": "Return",
    "return": "Return",
    "esc": "Escape",
    "escape": "Escape",
    "tab": "Tab",
    "space": "space",
    "backspace": "BackSpace",
    "delete": "Delete",
    "insert": "Insert",
    "up": "Up",
    "down": "Down",
    "left": "Left",
    "right": "Right",
    "home": "Home",
    "end": "End",
    "pageup": "Page_Up",
    "pagedown": "Page_Down",
    "ctrl": "ctrl",
    "control": "ctrl",
    "alt": "alt",
    "shift": "shift",
    "cmd": "super",
    "win": "super",
    "super": "super",
}


@dataclass(frozen=True)
class Screenshot:
    """A captured frame: PNG bytes plus the display dimensions."""

    data: bytes
    width: int
    height: int


@dataclass(frozen=True)
class DesktopDeps:
    """Internal deps from Sandbox. ``refresh_activate`` is invoked on 401."""

    sandbox_id: str
    sandbox_host: str
    get_access_token: Callable[[], str]
    refresh_activate: Callable[[], str]
    publish_stream_port: Callable[[], None]
    stream_base_url: Callable[[], str]


@dataclass(frozen=True)
class AsyncDesktopDeps:
    """Async variant of DesktopDeps."""

    sandbox_id: str
    sandbox_host: str
    get_access_token: Callable[[], str]
    refresh_activate: Callable[[], Awaitable[str]]
    publish_stream_port: Callable[[], Awaitable[None]]
    stream_base_url: Callable[[], str]


def _keysym(key: str) -> str:
    return _KEYSYMS.get(key.lower(), key)


def _chord_parts(key: str | Sequence[str]) -> tuple[list[str], str]:
    """Split a chord into (modifiers, final key) for the KeyEvent RPC.

    ``press("ctrl+c")`` and ``press(["ctrl", "c"])`` are equivalent.
    """
    parts = [_keysym(p) for p in (key.split("+") if isinstance(key, str) else key)]
    if not parts or any(p == "" for p in parts):
        raise ValueError("press: empty key")
    return parts[:-1], parts[-1]


def _assert_coordinate(x: int, y: int) -> None:
    # 0 is a valid coordinate — validate explicitly, never with truthiness.
    if not isinstance(x, int) or not isinstance(y, int) or x < 0 or y < 0:
        raise ValueError(f"Invalid coordinates ({x}, {y}): must be integers >= 0")


def _pointer_body(x: int, y: int, action: str, button: MouseButton = "left") -> dict:
    _assert_coordinate(x, y)
    return {"x": x, "y": y, "button": _BUTTONS[button], "action": action}


def _action_body(action: dict[str, Any]) -> dict[str, Any]:
    """Lower one ``desktop.actions()`` step into its proto-JSON shape."""
    kind = action.get("type")
    if kind == "click":
        return {
            "pointer": _pointer_body(
                action["x"],
                action["y"],
                "POINTER_ACTION_CLICK",
                action.get("button", "left"),
            )
        }
    if kind == "double_click":
        return {
            "pointer": _pointer_body(
                action["x"], action["y"], "POINTER_ACTION_DOUBLE_CLICK"
            )
        }
    if kind == "move":
        return {
            "pointer": _pointer_body(action["x"], action["y"], "POINTER_ACTION_MOVE")
        }
    if kind == "mouse_down":
        return {
            "pointer": _pointer_body(
                action["x"],
                action["y"],
                "POINTER_ACTION_DOWN",
                action.get("button", "left"),
            )
        }
    if kind == "mouse_up":
        return {
            "pointer": _pointer_body(
                action["x"],
                action["y"],
                "POINTER_ACTION_UP",
                action.get("button", "left"),
            )
        }
    if kind == "press":
        modifiers, final_key = _chord_parts(action["key"])
        return {"key": {"key": final_key, "modifiers": modifiers}}
    if kind == "write":
        return {"key": {"text": action["text"]}}
    if kind == "scroll":
        return {"scroll": {"dx": action.get("dx", 0), "dy": action.get("dy", 0)}}
    raise ValueError(f"Unknown desktop action type: {kind!r}")


def _decode_screenshot(raw: dict[str, Any]) -> Screenshot:
    image = raw.get("image")
    if image is None:
        raise ValueError("Screenshot response missing image data")
    return Screenshot(
        data=base64.b64decode(image),
        width=raw.get("width", 0),
        height=raw.get("height", 0),
    )


def _stream_url(base: str, *, view_only: bool) -> str:
    params: dict[str, str] = {"autoconnect": "1", "resize": "scale"}
    if view_only:
        params["view_only"] = "1"
    return f"{base}/vnc.html?{urlencode(params)}"


class Desktop:
    """Sync desktop control. Access as ``sandbox.desktop``."""

    def __init__(self, deps: DesktopDeps, client: httpx.Client | None = None) -> None:
        self._deps = deps
        self._client = client
        target = data_plane_target(deps.sandbox_id, deps.sandbox_host)
        self._data_plane_base_url = target.url
        self._routing_headers = target.headers

    def screenshot(self) -> Screenshot:
        """Capture the current screen as a PNG.

        One round trip; safe to call while a live stream viewer is open.
        """
        return _decode_screenshot(self._rpc("Screenshot", {}))

    def click(self, x: int, y: int, *, button: MouseButton = "left") -> None:
        """Click at (x, y). One RPC — move and click are a single action."""
        self._rpc("SendPointer", _pointer_body(x, y, "POINTER_ACTION_CLICK", button))

    def double_click(self, x: int, y: int) -> None:
        """Double-click at (x, y)."""
        self._rpc("SendPointer", _pointer_body(x, y, "POINTER_ACTION_DOUBLE_CLICK"))

    def right_click(self, x: int, y: int) -> None:
        """Right-click at (x, y)."""
        self.click(x, y, button="right")

    def middle_click(self, x: int, y: int) -> None:
        """Middle-click at (x, y)."""
        self.click(x, y, button="middle")

    def move_mouse(self, x: int, y: int) -> None:
        """Move the pointer to (x, y) without clicking."""
        self._rpc("SendPointer", _pointer_body(x, y, "POINTER_ACTION_MOVE"))

    def drag(
        self,
        start: tuple[int, int],
        end: tuple[int, int],
        *,
        button: MouseButton = "left",
    ) -> None:
        """Drag from ``start`` to ``end`` as one atomic batch (down-move-up)
        under the sandbox's input lock, so no other input can interleave."""
        self.actions(
            [
                {"type": "mouse_down", "x": start[0], "y": start[1], "button": button},
                {"type": "move", "x": end[0], "y": end[1]},
                {"type": "mouse_up", "x": end[0], "y": end[1], "button": button},
            ]
        )

    def scroll(self, *, dx: int = 0, dy: int = 0) -> None:
        """Scroll under the pointer. Positive ``dy`` scrolls down, negative
        up; ``dx`` scrolls horizontally. Both axes in one call."""
        self._rpc("Scroll", {"dx": dx, "dy": dy})

    def write(self, text: str) -> None:
        """Type literal text. Fast by default — no per-character pacing."""
        if not text:
            return
        self._rpc("SendKey", {"text": text})

    def press(self, key: str | Sequence[str]) -> None:
        """Press a key or chord: ``press("enter")``, ``press("ctrl+c")``,
        ``press(["ctrl", "shift", "p"])``. Friendly names map to X keysyms;
        unrecognized names pass through verbatim."""
        modifiers, final_key = _chord_parts(key)
        self._rpc("SendKey", {"key": final_key, "modifiers": modifiers})

    def actions(self, actions: Sequence[dict[str, Any]]) -> None:
        """Execute an ordered batch of actions in a single request.

        The whole batch is validated before anything runs, then executed
        under the sandbox's input lock. Execution stops at the first failing
        action. This is the fast path for models that emit several actions
        per turn.

        Example::

            sandbox.desktop.actions([
                {"type": "click", "x": 640, "y": 32},
                {"type": "write", "text": "https://example.com"},
                {"type": "press", "key": "enter"},
            ])
        """
        if not actions:
            return
        self._rpc("SendActions", {"actions": [_action_body(a) for a in actions]})

    def resize(self, width: int, height: int) -> None:
        """Resize the virtual display. Width must be a multiple of 8 between
        320 and 8192; height between 200 and 8192. Takes effect live."""
        self._rpc("Resize", {"width": width, "height": height})

    def get_stream_url(self, *, view_only: bool = False) -> str:
        """Publish the live desktop viewer (noVNC) and return its browser URL.

        The URL goes through the sandbox's preview-port access policy — under
        a private policy, viewers also need a preview token.
        """
        self._deps.publish_stream_port()
        return _stream_url(self._deps.stream_base_url(), view_only=view_only)

    def _rpc(self, method: str, body: dict[str, Any]) -> dict[str, Any]:
        def send(token: str) -> Any:
            return api_request(
                "POST",
                f"{self._data_plane_base_url}{_RPC_BASE}/{method}",
                headers={**self._routing_headers, "X-Access-Token": token},
                json_body=body,
                client=self._client,
            )

        return with_token_retry(
            self._deps.get_access_token, self._deps.refresh_activate, send
        )


class AsyncDesktop:
    """Async desktop control. Access as ``sandbox.desktop``."""

    def __init__(
        self, deps: AsyncDesktopDeps, client: httpx.AsyncClient | None = None
    ) -> None:
        self._deps = deps
        self._client = client
        target = data_plane_target(deps.sandbox_id, deps.sandbox_host)
        self._data_plane_base_url = target.url
        self._routing_headers = target.headers

    async def screenshot(self) -> Screenshot:
        """Async variant of :meth:`Desktop.screenshot`."""
        return _decode_screenshot(await self._rpc("Screenshot", {}))

    async def click(self, x: int, y: int, *, button: MouseButton = "left") -> None:
        """Async variant of :meth:`Desktop.click`."""
        await self._rpc(
            "SendPointer", _pointer_body(x, y, "POINTER_ACTION_CLICK", button)
        )

    async def double_click(self, x: int, y: int) -> None:
        """Async variant of :meth:`Desktop.double_click`."""
        await self._rpc(
            "SendPointer", _pointer_body(x, y, "POINTER_ACTION_DOUBLE_CLICK")
        )

    async def right_click(self, x: int, y: int) -> None:
        """Async variant of :meth:`Desktop.right_click`."""
        await self.click(x, y, button="right")

    async def middle_click(self, x: int, y: int) -> None:
        """Async variant of :meth:`Desktop.middle_click`."""
        await self.click(x, y, button="middle")

    async def move_mouse(self, x: int, y: int) -> None:
        """Async variant of :meth:`Desktop.move_mouse`."""
        await self._rpc("SendPointer", _pointer_body(x, y, "POINTER_ACTION_MOVE"))

    async def drag(
        self,
        start: tuple[int, int],
        end: tuple[int, int],
        *,
        button: MouseButton = "left",
    ) -> None:
        """Async variant of :meth:`Desktop.drag`."""
        await self.actions(
            [
                {"type": "mouse_down", "x": start[0], "y": start[1], "button": button},
                {"type": "move", "x": end[0], "y": end[1]},
                {"type": "mouse_up", "x": end[0], "y": end[1], "button": button},
            ]
        )

    async def scroll(self, *, dx: int = 0, dy: int = 0) -> None:
        """Async variant of :meth:`Desktop.scroll`."""
        await self._rpc("Scroll", {"dx": dx, "dy": dy})

    async def write(self, text: str) -> None:
        """Async variant of :meth:`Desktop.write`."""
        if not text:
            return
        await self._rpc("SendKey", {"text": text})

    async def press(self, key: str | Sequence[str]) -> None:
        """Async variant of :meth:`Desktop.press`."""
        modifiers, final_key = _chord_parts(key)
        await self._rpc("SendKey", {"key": final_key, "modifiers": modifiers})

    async def actions(self, actions: Sequence[dict[str, Any]]) -> None:
        """Async variant of :meth:`Desktop.actions`."""
        if not actions:
            return
        await self._rpc("SendActions", {"actions": [_action_body(a) for a in actions]})

    async def resize(self, width: int, height: int) -> None:
        """Async variant of :meth:`Desktop.resize`."""
        await self._rpc("Resize", {"width": width, "height": height})

    async def get_stream_url(self, *, view_only: bool = False) -> str:
        """Async variant of :meth:`Desktop.get_stream_url`."""
        await self._deps.publish_stream_port()
        return _stream_url(self._deps.stream_base_url(), view_only=view_only)

    async def _rpc(self, method: str, body: dict[str, Any]) -> dict[str, Any]:
        async def send(token: str) -> Any:
            return await async_api_request(
                "POST",
                f"{self._data_plane_base_url}{_RPC_BASE}/{method}",
                headers={**self._routing_headers, "X-Access-Token": token},
                json_body=body,
                client=self._client,
            )

        return await async_with_token_retry(
            self._deps.get_access_token, self._deps.refresh_activate, send
        )
